pyro/contrib/gp/models/sgpr.py

```diff
@@ -83,14 +81,14 @@ def model(self):
             else:  # approx = "VFE"
                 trace_term += (Kffdiag - Qffdiag).sum() / noise
 
-        # correct event_shape for y
-        y_t = self.y.t() if self.y.dim() == 2 else self.y
-        zero_loc = y_t.new([0]).expand(y_t.size())
         # DTC: cov = Qff + noise, trace_term = 0
         # FITC: cov = Qff + diag(Kff - Qff) + noise, trace_term = 0
         # VFE: cov = Qff + noise, trace_term = tr(Kff - Qff) / noise
-        pyro.sample("y", dist.SparseMultivariateNormal(zero_loc, D, W, trace_term).reshape(
-            extra_event_dims=zero_loc.dim() - 1), obs=y_t)
+        # convert y_shape from N x D to D x N
```

Comment:
I'm surprised this is needed, since we generally try to keep batch dimensions on the left. Can you explain a bit more what issues you were seeing? (Feel free to ping me on PyTorch slack, or we could talk over Zoom)

Reply:
Here `N` is event_shape, and `D` is batch_shape (but people traditionally set N x D for y, I don't want to break that). This swap is to guarantee y_shape follows Pyro batch semantics: batch_shape x event_shape.